app/components/Settings/Preview/code-editor.jsx

```diff
@@ -0,0 +1,232 @@
+     */
+    constructor(props) {
+        super(props);
+
+        /**
+         * @member {object} state - Component state
+         * @property {object} state.tables - Tables schemas (used in hintOptions)
+         * @property {number} state.height - Height in pixels
+         * @property {?number} state.width - Width in pixels
+         */
+        this.state = {
+            height: 250,
+            width: null,
+            minConstraints: [200, 74],
+            maxConstraints: [Infinity, Infinity],
+            tables: CodeEditor.computeAutocompleteTables(this.props)
+        };
+
+        this.onBeforeChange = (editor, data, value) => {
+            this.props.onChange(value);
+        };
+
+        this.onChange = (editor) => {
+            // Don't show autocomplete after a space or semicolon
+            const ch = editor.getTokenAt(editor.getCursor()).string.slice(-1);
+            if (!ch || ch === ' ' || ch === ';') {
+                return;
+            }
+
+            // hint options for specific plugin & general show-hint
+            // 'tables' is sql-hint specific
+            // 'disableKeywords' is also sql-hint specific, and undocumented but referenced in sql-hint plugin
+            // Other general hint config, like 'completeSingle' and 'completeOnSingleClick'
+            // should be specified here and will be honored
+            const hintOptions = {
+                tables: this.state.tables,
+                completeSingle: false,
+                completeOnSingleClick: true
+            };
+
+            // editor.hint.sql is defined when importing codemirror/addon/hint/sql-hint
+            // (this is mentioned in codemirror addon documentation)
+            // Reference the hint function imported here when including other hint addons
+            // or supply your own
+            CM.showHint(editor, CM.hint.sql, hintOptions);
+        };
+
+        this.editorDidMount = (editor) => {
+            this.editor = editor;
+
+            const wrapperElement = this.editor.getWrapperElement();
+            const height = wrapperElement.clientHeight;
+            const width = wrapperElement.clientWidth;
+
+            const minConstraints = [width, 74];
+            const maxConstraints = [width, Infinity];
+
+            this.setState({height, width, minConstraints, maxConstraints});
+        };
+
+        this.onResize = (event, data) => {
+            const height = data.size.height;
+
+            const wrapperElement = this.editor ? this.editor.getWrapperElement() : {};
+            const width = wrapperElement.clientWidth;
+
+            if (this.editor) this.editor.setSize(width, height);
+
+            this.setState({height, width});
+        };
+    }
+
+    static computeAutocompleteTables(props) {
+        const schemaRequest = props.schemaRequest || {};
+
+        if (schemaRequest.status !== 200 || !schemaRequest.content || !schemaRequest.content.rows) {
```

Comment:
Should we be moving connection status to a centralized location or atleast have this as a const?  Minor thing

Reply:
No, let's not do that. There is no gain in having `STATUS_OK_200` vs  `200`.